src/Components/Components/src/RenderTree/EventHandlerDelegateWithClosuresComparer.cs

```diff
@@ -0,0 +1,130 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Linq;
+using System.Diagnostics.CodeAnalysis;
+using System.Runtime.CompilerServices;
+using System.Collections.Concurrent;
+using System.Reflection;
+using System.Linq.Expressions;
+
+namespace Microsoft.AspNetCore.Components.RenderTree
+{
+    internal static class EventHandlerDelegateWithClosuresComparer
+    {
+        private static readonly ConcurrentDictionary<Type, Func<object, object, bool>> ClosureComparerCache = new();
+
+        internal static bool EventCallBackEquals(ref EventCallback left, ref EventCallback right)
+        {
+            if (Equals(left, right))
+            {
+                return true;
+            }
+
+            // Whenever the receivers are different and they are explicit we assume that
+            // the EventCallbacks are not equal.
+            if (left.Receiver != right.Receiver &&
+                (left.RequiresExplicitReceiver || right.RequiresExplicitReceiver))
+            {
+                return false;
+            }
+
+            return DelegateEquals(left.Delegate, right.Delegate);
+        }
+
+        internal static bool DelegateEquals(MulticastDelegate? left, MulticastDelegate? right)
+        {
+            if (Equals(left, right))
+            {
+                return true;
+            }
+
+            // If any of the delegates or their targets are null (static delegates) then normal
+            // equality is sufficient. No additional testing required.
+            if (left == null || right == null || left.Target == null || right.Target == null)
+            {
+                return false;
+            }
+
+            var oldTargetType = left.Target!.GetType();
+            var newTargetType = right.Target!.GetType();
+
+            // if the types are not the same, or the methods are not the same, these possible closures are
+            // not the same anyway.
+            if (oldTargetType != newTargetType || left.Method != right.Method)
+            {
+                return false;
+            }
+
+            // for speed we cache the comparison functions per closure type.
+            var comparison = ClosureComparerCache.GetOrAdd(newTargetType, GetClosureComparerForType);
```

Comment:
I'm curious about `newTargetType` being the dictionary key here rather than the `MethodInfo`.

I'm guessing that, currently, the C# compiler produces a separate type to hold the captured variables for each closure and doesn't reuse them even if the captures variables happened to be the same. But can we be sure that will be true forever? Wouldn't it be safer to key this from the `MethodInfo` - doesn't that implicitly tell you what the target type is anyway?

Reply:
Well, the comparison function would be the same for the closure, even if the delegates (or nested delegates) eventually would return false in comparison because of the fact that they point to different methods. So you really can cache per type, no need to cache per method info. The "way" to compare does not change.

Reply:
Makes sense - thanks!